Writing Guidelines: quicklinks and surfacing content (#31248)
mdn/content: files/en-us/mdn/writing_guidelines/page_structures/quicklinks/index.md

---
title: Sidebars and other link macros
slug: MDN/Writing_guidelines/Page_structures/Quicklinks
page-type: mdn-writing-guide
---

{{MDNSidebar}}

> MDN supports adding links and lists of links in sidebars and within the main content of a page. You can and often **should** use macros to generate links.

Reviewer: "can and often should" is not helpful advice without explaining somewhere what "often" means. IMO you _should_ use macros for sidebars, and for xrefs when you are linking to a reference page for a concrete web platform feature (and an xref for that feature exists).

This article describes how to use the different MDN macros, demonstrating how to include sidebars to MDN pages and add always-up-to-date and maintained links within your content. Quickly.

In this guide, you will learn how to create a sidebar of links by just including a macro and how to add how to use those same macros to create sidebars with additional text. You will also learn about cross-reference macros to add links to other pages and macros to add lists of links to pages and subpages.

## Creating sidebars

> Every page has a sidebar. These sidebars are created by Quicklinks. **Quicklinks** are MDN macros that create boxes containing a potentially hierarchical list of links to other pages on this site. They are mainly used in sidebars, though can also be used in the main content area.

Reviewer: Like Hamish says, the term "quicklinks" doesn't add anything here.

When a Quicklink is included, the server creates a section of content containing an unordered list of links. The links created, where they are displayed, and how they are displayed depend on the macro used and the parameters included in the markdown macro call. Some quicklinks include links based on a directory's structure or page type. Others include a list of predefined pages that are hard-coded in Yari.

### Fully macro generated sidebars

To include only the content generated by the Quicklink macro, the macro is added immediately after the frontmatter and before the content on every page. For example, the first lines of this document are written as:

```md
---
title: Sidebars and other link macros
slug: MDN/Writing_guidelines/Page_structures/Quicklinks
page-type: mdn-writing-guide
---

\{{MDNSidebar}}
```

The `\{{MDNSidebar}}` is a Quicklink macro that adds the MDN sidebar to the page. Because we only want the content from the macro, and no additional content in our sidebar, the macro is placed immediately after the frontmatter.

Here are a few other Quicklink macros, with what they do:

- `\{{CSSRef}}`

>   - : Present on every CSS page, it generates different CSS sidebars depending on tags, like "properties", "selectors", "at-rules", "descriptors", etc., included in the page.

Reviewer: No, it always generates the same sidebar.

> - `\{{ListSubpagesForSidebar(<parameters>)}}`

Reviewer: ListSubpagesForSidebar does not make a sidebar, it makes an ordered list of links, that can be included in a sidebar. (that is, it does not make a section with a `quick_links` ID, which is the magic that makes a blob of HTML get treated as a sidebar by Yari).

  - : Inserts a tree of subpages of the slug of the pages specified as the first parameter. Include two parameters, with the second being `true` or `1`, to display the links as plain text instead of like code. Include a third parameter, set to `true` or `1`, to include the parent page at the top of the list with the link text "Overview".

- `\{{DefaultAPISidebar("<API_Title>")}}`

  - : The API sidebar displayed for overview pages; the single parameter is the name of the API group in GroupData.

- Other sidebar "quicklink" macros
  - : Other sidebar generating macro calls include `\{{GlossarySidebar}}`, `\{{LearnSidebar}}`,`\{{HTMLSidebar}}`, `\{{HTTPSidebar}}`, `\{{PWASidebar}}`, etc.

The appropriate macro to use depends on the page type and is listed in the template for each page type.

### Sidebars: adding additional content

Each of the Quicklink macros listed above creates a predefined sidebar structure.

> If you need to append links before or after the sidebar structure generated by the macro, you can do so by:

Reviewer: I think this is bad advice. A given page in a given bit of the site should have a given sidebar - so e.g. a page in CSS should have CSSRef. All pages in that bit of the site should have the same sidebar.

If you want to add links to a page in CSS, you should want to add links to all pages in CSS. So you should handle this by modifying the sidebar macro.

In general, if we want to document "manual quicklinks" at all, we should document them as an anti-pattern that people will sometimes find in parts of the site that don't yet have sidebar macros defined for them.

## Main content area links

Macros are not limited to sidebars or even to long lists of links.

Quicklink macros can also be used to add lists of links to the main area of the page.

- `\{{LandingPageListSubPages()}}`
  - : Inserts a definition list ({{HTMLelement("dl")}}) of the subpages of the current page, with each page's title as the {{HTMLelement("dt")}} term and its SEO summary as the {{HTMLelement("dd")}} term. The optional parameter accepts the slug or the parent page of the directory of pages to output instead of the subpages of the current page.

### Cross-reference links

In addition to macros that create lists of links, there are macros that quickly create a single link to cross-reference a CSS, JavaScript, SVG, or HTML feature, including attributes, elements, properties, data types, and APIs. The macros that create single links require at least one parameter: the feature being referenced.

The macros include:

- `\{{CSSxRef("")}}`
- `\{{DOMxRef("")}}`
- `\{{HTMLElement("")}}`
- `\{{JSxRef("")}}`
- `\{{SVGAttr("")}}`
- `\{{SVGElement("")}}`
- `\{{HTTPHeader("")}}`
- `\{{HTTPMethod("")}}`
- `\{{HTTPStatus("")}}`

> All the macros accept additional parameters. The second parameter, if present, provides the link text. By default, these links are displayed like code, in a monospace font. To prevent HTML code semantics and CSS code styling, the `"nocode"` parameter can be included as the third value in the macro call.

Reviewer: Unfortunately, the different xref macros don't all have the same API, so it's hard to make general statements about them. For example, CSSXref doesn't support "nocode" at all, while for those that do, like DOMXref, it is more often the fourth parameter than the third. I think it would be better either to document the parameters completely, or just to refer to the source files, which should always include documentation: https://github.com/mdn/yari/blob/bda844f7dd745f7e3ac15736c3c884fc26507464/kumascript/macros/DOMxRef.ejs#L1-L15.

> For example, `\{{CSSxRef("background-color")}}` creates the code link "{{CSSxRef("background-color")}}" and `\{{CSSxRef("At-rules", "CSS at-rule", "nocode")}}` creates the plain text link "{{CSSxRef("At-rules", "CSS at-rule", "nocode")}}".

Reviewer: https://developer.mozilla.org/en-US/docs/Web/CSS/At-rules doesn't exist. Also I think it is an anti-pattern to use xefs to link to guide pages, although we are not very clear about this.

Here are a few other quicklink macros, with what they do:

- `\{{CSSRef}}`

  - : Present on every CSS page, it generates different CSS sidebars depending on tags, like "properties", "selectors", "at-rules", "descriptors", etc., included in the page.

- `\{{ListSubpagesForSidebar(<parameters>)}}`

  - : Inserts a tree of subpages of the slug of the pages specified as the first parameter. Include two parameters, with the second being `true` or `1`, to display the links as plain text instead of like code. Include a third parameter, set to `true` or `1`, to include the parent page at the top of the list with the link text "Overview".

- `\{{DefaultAPISidebar("<API_Title>")}}`

  - : The API sidebar displayed for overview pages; the single parameter is the name of the API group in GroupData.

- Other sidebar "quicklink" macros
  - : Other sidebar generating macro calls include `\{{GlossarySidebar}}`, `\{{LearnSidebar}}`,`\{{HTMLSidebar}}`, `\{{HTTPSidebar}}`, `\{{PWASidebar}}`, etc.

The appropriate macro to use depends on the page type and is listed in the template for each page type.

> ### Adding content to your sidebar before or after the quicklink
> 
> Each macro listed above creates sections of links displayed automatically in the sidebar. No additional code is needed to define the location of the list. The macro call needs to be nested within specific HTML to have the list of links add additional content included as a sidebar.
> 
> To include a sidebar with additional content — to display content above or below the list of links outputted by a macro — the quicklinks macro must be appended to the end of the page (instead of between the frontmatter and the page content) with the macro call nested in a {{htmlelement("section")}} element with `id="Quick_links"` set.
> 
> For example, adding the following at the end of a markdown file will create a sidebar containing the links to all the ARIA role pages, preceded by a link to the ARIA roles overview page:
> 

Reviewer 2: This is a proposal that attempts to overcome my preconceptions about the meaning of the term quicklinks macro vs sidebar macro.

Happy for you to amend.

```suggestion
### Adding content to your sidebar before or after the quicklink

Each of the quicklink macros listed above creates a predefined sidebar structure.
If you need to append links before or after the sidebar structure generated by the macro, you can do so by nesting both the macro and any additional links within an HTML {{htmlelement("section")}} element that has the `id="Quick_links"`.
This section must be appended to the end of the page, instead of between the frontmatter and the page content, and any sidebar quicklink macro below the frontmatter must be removed (there can only be one sidebar defined in each document).

For example, adding the following at the end of a markdown file (and removing any sidebar macro from below the frontmatter, will create a sidebar containing the links to all the ARIA role pages, preceded by a link to the ARIA roles overview page:
```

```md
<section id="Quick_links">

1. [**WAI-ARIA roles**](/en-US/docs/Web/Accessibility/ARIA/Roles)

   \{{ListSubpagesForSidebar("/en-US/docs/Web/Accessibility/ARIA/Roles", "true")}}

</section>
```

> > **Note:** To include content in your sidebar along with a quicklink macro, the extra content and the quicklinks macro must be placed at the end of the document. Only one sidebar is created per page, so any macro listed after the frontmatter must be removed.

Reviewer 2: Worth making it clear you don't have to use a macro at all? Also that the items you add should be in an ordered list?

```suggestion

You can also fully craft a sidebar without using any quicklink macro.
Note however that links that you add that aren't generated by a quicklinks macro should be inserted as an ordered markdown list.

> **Note:** To include content in your sidebar along with a quicklink macro, the extra content and the quicklinks macro must be placed at the end of the document. Only one sidebar is created per page, so any macro listed after the frontmatter must be removed.
```

- `\{{DefaultAPISidebar("<API_Title>")}}`

  - : The API sidebar displayed for overview pages; the single parameter is the name of the API group in GroupData.

- Other sidebar "quicklink" macros
  - : Other sidebar generating macro calls include `\{{GlossarySidebar}}`, `\{{LearnSidebar}}`,`\{{HTMLSidebar}}`, `\{{HTTPSidebar}}`, `\{{PWASidebar}}`, etc.

The appropriate macro to use depends on the page type and is listed in the template for each page type.

### Adding content to your sidebar before or after the quicklink

Each of the quicklink macros listed above creates a predefined sidebar structure.
If you need to append links before or after the sidebar structure generated by the macro, you can do so by nesting both the macro and any additional links within an HTML {{htmlelement("section")}} element that has the `id="Quick_links"`.
This section must be appended to the end of the page, instead of between the frontmatter and the page content, and any sidebar quicklink macro below the frontmatter must be removed (there can only be one sidebar defined in each document).

> For example, adding the following at the end of a markdown file (and removing any sidebar macro from below the frontmatter, will create a sidebar containing the links to all the ARIA role pages, preceded by a link to the ARIA roles overview page:

Reviewer 2: ```suggestion
For example, adding the following at the end of a markdown file (and removing any sidebar macro from below the frontmatter), will create a sidebar containing the links to all the ARIA role pages, preceded by a link to the ARIA roles overview page:
```

> {{MDNSidebar}}

Reviewer 2: Both this and the original page IMO overcomplicate things. For end users Quicklinks is _just_ the mechanism to append your own set of links to a sidebar rather than/in addition to the crafted sidebar.

What both this and the original do is mix that simple concept with all sorts of stuff about how sidebars are generated and how you can generate your own lists, and some examples of how the underlying system works. None of that is particularly necessary - in particular because this feature is very much something not to be recommended.

If it were me I'd revert most of this except the changes you did to the table - which are great. And at the end of that append a simple markdown ordered list example - because in most cases that is how it is really done by the consumers of this document.

Author: Quicklinks was a bad page, so it needed updating, so I prefer not to revert. Instead, I removed the stuff that really shouldn't be there, added some see alsos, and added the compat and spec macro to the page structure macros page where they were missing and its a better fit.

Reviewer 2: OK, so I _think_ I may understand the problem. 

To me Quicklinks is **entirely** the trick of using `<section id="Quick_links">INSERT ORDERED LIST</section>` to add a custom sidebar. It's an occasional trick, where you might use a macro to generate the list of things, and indeed some of the the standard sidebars do this.

Looking this new doc, you think Quicklinks is the practice of auto generating links via macros in either content and sidebars. Because of that this goes on about the various macros and sidebars.

So now I don't know what to say - firstly it depends on how you define Quicklinks and what you expect people to do with this document. Secondly, you could argue that if this is about Sidebars, then the topic should be named sidebars, and focused on how sidebars work. We really don't use those list macros outside of sidebars except in very specific  cases and I'm not sure what users now get out of this.

Hope that makes sense. Perhaps a second opinion on this? Will or Jean-Yves.

- `\{{ListSubpagesForSidebar(<parameters>)}}`

  - : Inserts a tree of subpages of the slug of the pages specified as the first parameter. Include two parameters, with the second being `true` or `1`, to display the links as plain text instead of like code. Include a third parameter, set to `true` or `1`, to include the parent page at the top of the list with the link text "Overview".

- `\{{DefaultAPISidebar("<API_Title>")}}`

  - : The API sidebar displayed for overview pages; the single parameter is the name of the API group in GroupData.

- Other sidebar macros
  - : Other sidebar generating macro calls include `\{{GlossarySidebar}}`, `\{{LearnSidebar}}`,`\{{HTMLSidebar}}`, `\{{HTTPSidebar}}`, `\{{PWASidebar}}`, etc.

The appropriate macro to use depends on the page type and is listed in the template for each page type.

### Adding static content to your sidebar quicklinks

Each macro listed above creates sections of links displayed automatically in the sidebar. No additional code is needed to define the location of the list. The macro call needs to be nested within specific HTML to have the list of links add additional content included as a sidebar.

> To include a sidebar with additional content — to display content above or below the list of links outputted by a macro — the quicklinks macro must be appended to the end of the page (instead of between the frontmatter and the page content) with the macro call nested in a {{htmlelement("section")}} element with `id="Quick_links"` set.

Reviewer 2: I just tested this by adding some quicklinks to the end of https://developer.mozilla.org/en-US/docs/Web/API/Screen_Wake_Lock_API - the links were not appended.

However when I deleted the default sidebar from below the frontmatter the quicklinks were displayed.

Can you sanity check this too?

Author: yes. You need to delete the macro that comes between the frontmatter and the start of hte content for the macro in the `<section id="qL">{{macro}} + text</section>` to work.

I added a comment to make it clearer

#### Quick links: macros to create single links

In addition to macros that create lists of links, there are macros that quickly create a single link for a CSS, JavaScript, SVG, or HTML feature, including attributes, elements, properties, data types, and APIs. The macros that create single links require at least one parameter: the feature being referenced.

The macros include:

- `\{{CSSxRef("")}}`
- `\{{DOMxRef("")}}`
- `\{{HTMLElement("")}}`
- `\{{JSxRef("")}}`
- `\{{SVGAttr("")}}`
- `\{{SVGElement("")}}`
- `\{{HTTPHeader("")}}`
- `\{{HTTPMethod("")}}`
- `\{{HTTPStatus("")}}`

All the macros accept additional parameters. The second parameter, if present, provides the link text. By default, these links are displayed like code, in a monospace font. To prevent HTML code semantics and CSS code styling, the `"nocode"` parameter can be included as the third value in the macro call.

> For example, `\{{CSSxRef("background-color")}}` creates the code link "{{CSSxRef("background-color")}}" and `\{{CSSxRef("at-rule", "CSS at-rule", "nocode")}}` creates the plain text link "{{CSSxRef("at-rule", "CSS at-rule", "nocode")}}".

Reviewer 2: ```suggestion
For example, `\{{CSSxRef("background-color")}}` creates the code link "{{CSSxRef("background-color")}}" and `\{{CSSxRef("At-rules", "CSS at-rule", "nocode")}}` creates the plain text link "{{CSSxRef("At-rules", "CSS at-rule", "nocode")}}".
```

  - : Creates a set of quicklinks using the current page's (or the specified page's) children as the destinations.
    This creates hierarchical lists up to two levels deep.
    The pages' titles are used as the link text and their summaries as tooltips.

### Additional macros

In addition to quicklinks, there are a several other macros that you can include.

#### Page sections

The following macros are included on all reference pages, but are also supported by all page types:

- `\{{Compat}}` / `\{{Compat(&lt;feature>)}}` / `\{{Compat(&lt;feature>, &lt;depth>)}}`

  - : Generates a [compatibility table](/en-US/docs/MDN/Writing_guidelines/Page_structures/Compatibility_tables) for the feature passed as the parameter. If no parameter is included, it defaults to the features defined by `browser-compat` in the frontmatter. An optional depth parameter sets how deep sub features should be added to the table. The depth, if omitted, defaults to 1, meaning only the first level of sub feature data from BCD will be included.

- `\{{Specifications}}` / `\{{Specifications(&lt;feature>)}}`

>   - : Includes the specification for the feature specified in the parameter. If no parameter is passed, the specification listed is defined by the value for `spec_urls` in the frontmatter, if present, or from the specification listed in browser compatibility data defined by `browser-compat` in the frontmatter. The specification is rendered as an external link.

Reviewer 2: Ditto, none of these should have any definition in this page, but are reasonably cross linked from macros page.

### Banner macros

- `\{{SeeCompatTable}}`

  - : Generates a **This is an experimental technology** banner indicating the technology is [experimental](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#experimental) with a link to the [Browser compatibility](/en-US/docs/MDN/Writing_guidelines/Page_structures/Compatibility_tables) that every [reference page](/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types) contains.

- `\{{Deprecated_Header}}`

  - : Generates a **Deprecated** banner indicating the technology is [deprecated](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#deprecated) and should not be used.

- `\{{SecureContext_Header}}`
  - : Generates a **Secure context** banner indicating a [secure context](/en-US/docs/Web/Security/Secure_Contexts) is required for the feature.

> For more information, see the [banners and notices guide](/en-US/docs/MDN/Writing_guidelines/Page_structures/Banners_and_notices)

Reviewer 2: None of these have anything to do with Quicklinks. Suggest remove but perhaps add to see also

Author: moved to a link to see also to the new location in the common macros page.
